modules/consensus/accept.go

```diff
@@ -140,20 +210,31 @@ func (cs *ConsensusSet) managedAcceptBlock(b types.Block) error {
 	err := cs.db.View(func(tx *bolt.Tx) error {
 		// Do not accept a block if the database is inconsistent.
 		if inconsistencyDetected(tx) {
-			return errors.New("inconsistent database")
+			return errInconsistentSet
 		}
 
-		// Check that the header is valid. The header is checked first because it
-		// is not computationally expensive to verify, but it is computationally
-		// expensive to create.
-		err := cs.validateHeader(boltTxWrapper{tx}, b)
+		// Do some relatively inexpensive checks to validate the header and block.
+		// Validation generally occurs in the order of least expensive validation
+		// first.
+		err := cs.validateHeaderAndBlock(boltTxWrapper{tx}, b)
 		if err != nil {
 			// If the block is in the near future, but too far to be acceptable, then
 			// save the block and add it to the consensus set after it is no longer
 			// too far in the future.
+			//
+			// TODO: an attacker could mine many blocks off the genesis block all in the
+			// future and we would spawn a goroutine per each block. To fix this, either
+			// ban peers that send lots of future blocks OR stop spawning goroutines
+			// after we are already waiting on a large number of future blocks.
```

Comment:
probably both. You want a hard-limit on all resources that you use, I'd never advocate having a system that can spin up unlimited goroutines.